### PR TITLE
Update cross chain commands - Closes #7575

### LINKED ISSUE #7575: Update cross chain commands
### Description

- Remove `CCForwardCommand`
- Update `CrossChainTransferCommand` as specified in the [PR](https://github.com/LiskHQ/lips/blob/08935fbff959dd14f1fc0c2b1b64890f904d5eae/proposals/lip-0051.md#cross-chain-token-transfer-messages)

### Acceptance Criteria

- CCForwardCommand src and tests are removed
- Updated CrossChainTransferCommand is unit tested

### Additional Information
- This change requires #7571 and #7572

### PR DESCRIPTION
**Pull request by mosmartin**
### What was the problem?

This PR resolves #7575

### How was it solved?

- Add verify() method
- Refactor execute() method
- Update the schema & constants

### How was it tested?

- Unit tests WIP

### CODE REVIEW

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 1]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
-		this._minBalances = args.minBalances;
+		// this._minBalances = args.minBalances;
 	}
 
-	public async execute(ctx: CCCommandExecuteContext): Promise<void> {
-		const { ccm } = ctx;
+	public async verify(ctx: CCCommandExecuteContext): Promise<void> {
+		const { ccm, ccmSize } = ctx;
 		const methodContext = ctx.getMethodContext();
 		const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:62]: ownChainID can be obtained from `init` function passed from module [here](https://github.com/LiskHQ/lisk-sdk/blob/feature/7211-update_token_module/framework/src/modules/token/module.ts#L159)

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 2]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
 
 	public init(args: { minBalances: MinBalance[]; supportedTokenIDs: Buffer[] }): void {
 		this._supportedTokenIDs = args.supportedTokenIDs;
-		this._minBalances = args.minBalances;
+		// this._minBalances = args.minBalances;
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:56]: both of these can be removed

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 3]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
+		const { ccm, ccmSize } = ctx;
 		const methodContext = ctx.getMethodContext();
 		const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
 		let params: CCTransferMessageParams;
-		let tokenChainID;
-		let tokenLocalID;
+		let tokenChainID: Buffer;
+		const { sendingChainID, status, fee } = ccm;
+
 		try {
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:67]: I think with the new definition, this try-catch is not needed?

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 4]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 			params = codec.decode<CCTransferMessageParams>(crossChainTransferMessageParams, ccm.params);
 			validator.validate(crossChainTransferMessageParams, params);
+			const { tokenID, amount } = params;
+			[tokenChainID] = splitTokenID(tokenID);
+
+			if (
+				!tokenChainID.equals(ownChainID) &&
+				!tokenChainID.equals(sendingChainID) &&
+				!tokenChainID.equals(tokenChainID)
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:76]: this should be mainchainID?
mainchainID can be obtained using
```
this._tokenMethod.getMainchainTokenID();
```

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 5]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
-					throw new Error(
-						`Amount ${params.amount.toString()} is not sufficient for ${escrowedAmount.toString()}`,
-					);
+
+				if (escrowedAmount < amount) {
+					throw new Error('Insufficient balance in escrow account.');
 				}
 			}
 		} catch (error) {
 			ctx.logger.debug({ err: error as Error }, 'Error verifying the params.');
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:95]: i think termination doesn't exist, so this catch is not needed

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 6]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 
-			return;
+			await this._interopMethod.terminateChain(methodContext, sendingChainID);
 		}
+	}
+
+	public async execute(ctx: CCCommandExecuteContext): Promise<void> {
+		const { ccm, ccmSize } = ctx;
+		const methodContext = ctx.getMethodContext();
+		const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:108]: same as above, ownChainID can be injected in `init`

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 7]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
+				await this._interopMethod.error(methodContext, ccm, CCM_STATUS_PROTOCOL_VIOLATION);
+			}
 
-		if (
-			!tokenSupported(this._supportedTokenIDs, params.tokenID) &&
-			ccm.fee >= ctx.ccmSize * MIN_RETURN_FEE &&
-			ccm.status === CCM_STATUS_OK
-		) {
-			await this._interopMethod.error(methodContext, ccm, CCM_STATUS_TOKEN_NOT_SUPPORTED);
+			await this._interopMethod.terminateChain(methodContext, sendingChainID);
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:123]: these logic doesn't exist anymore, so safe to do without try-catch

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 8]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 			return;
 		}
+		const { tokenID, amount, senderAddress } = params;
+		recipientAddress = params.recipientAddress;
+		const [tokenChainID, tokenLocalID] = splitTokenID(tokenID);
 
-		let { recipientAddress } = params;
-		if (ccm.status !== CCM_STATUS_OK) {
-			recipientAddress = params.senderAddress;
+		if (!tokenSupported(this._supportedTokenIDs, tokenID)) {
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:130]: `tokenSupported`, you can do with
```
const supported = await this.stores.get(SupportedTokensStore).isSupported(context, tokenID);
```

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 9]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
+			// TODO: emit event else throw error
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:131]: error, you can emit like
```
this.events.get(CcmTransferEvent).error(context, { ... }, TokenEventResult.TOKEN_NOT_SUPPORTED);
throw new Error(`tokenID ${tokenID.toStrong('hex')} is not supported`);
```

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 10]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
-				const supplyStore = this.stores.get(SupplyStore);
-				const supply = await supplyStore.get(methodContext, canonicalLocalID);
-				supply.totalSupply -= minBalance;
-				await supplyStore.set(methodContext, canonicalLocalID, supply);
+		if (status !== CCM_STATUS_OK) {
+			recipientAddress = senderAddress;
+		}
+
+		const userStore = this.stores.get(UserStore);
+		const user = await userStore.get(methodContext, userStore.getKey(recipientAddress, tokenID));
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:139]: `get` will throw with `NotFoundError`, so if you want to handle non existing case, you need to handle like below
```
try {
  const user = await userStore.get(methodContext, userStore.getKey(recipientAddress, tokenID));
} catch (error) {
  if (!(error instanceof NotFoundError)) {
    throw error
  }
  // do not found error case
}
```
but maybe
```
const recipientExist = await userStore.has(methodContext, userStore.getKey(recipientAddress, tokenID));
```
is sufficient in this case

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 11]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
+		if (!user) {
+			if (fee < FEE_CCM_INIT_USER_STORE) {
+				throw new Error('Insufficient fee to initialize user account.');
+			} else {
+				// TODO: if the relayer does not have enough balance, burn will raise exception
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:144]: you can use `this._tokenMethod.burn(methodContext, ...)`

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 12]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 			}
 		}
 
 		if (tokenChainID.equals(ownChainID)) {
 			const escrowStore = this.stores.get(EscrowStore);
-			const escrowKey = Buffer.concat([ccm.sendingChainID, tokenLocalID]);
+			const escrowKey = Buffer.concat([sendingChainID, tokenLocalID]);
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:150]:
```suggestion
			const escrowKey = escrowStore.getKey(sendingChainID, tokenID);
```

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 13]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 			const escrowData = await escrowStore.get(methodContext, escrowKey);
 
-			escrowData.amount -= params.amount;
+			escrowData.amount -= amount;
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:153]: I think we should check if `escrowData.amount > amount`

File: framework/src/modules/token/cc_commands/cc_transfer.ts [hunk 14]
```diff
@@ -56,99 +53,109 @@ export class CCTransferCommand extends BaseCCCommand {
...
 		}
 
-		const userStore = this.stores.get(UserStore);
-		await userStore.addAvailableBalanceWithCreate(
-			methodContext,
-			recipientAddress,
-			canonicalTokenID,
-			receivedAmount,
-		);
+		await userStore.addAvailableBalanceWithCreate(methodContext, recipientAddress, tokenID, amount);
```
Comment (shuse2) [inline framework/src/modules/token/cc_commands/cc_transfer.ts:157]: `withCreate` is probably not needed here

File: framework/src/modules/token/schemas.ts [hunk 15]
```diff
@@ -226,29 +227,27 @@ export const crossChainTransferMessageParams = {
 	properties: {
 		tokenID: {
 			dataType: 'bytes',
+			length: TOKEN_ID_LENGTH,
```
Comment (shuse2) [inline framework/src/modules/token/schemas.ts:230]: `length` actually doesn't exist on schema, so in the implementation `minLength` and `maxLength` are correct

File: framework/src/modules/token/schemas.ts [hunk 16]
```diff
@@ -226,29 +227,27 @@ export const crossChainTransferMessageParams = {
...
 		},
 		senderAddress: {
 			dataType: 'bytes',
+			length: ADDRESS_LENGTH,
 			fieldNumber: 3,
-			format: 'lisk32',
 		},
 		recipientAddress: {
 			dataType: 'bytes',
+			length: ADDRESS_LENGTH,
```
Comment (shuse2) [inline framework/src/modules/token/schemas.ts:244]: for `address`, `format: Lisk32` will check the length for dataType: `bytes`

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/src/modules/token/cc_commands/cc_transfer.ts (hunk 3 above)
Comment (mosmartin) [reply framework/src/modules/token/cc_commands/cc_transfer.ts:67]: I think we need it to verify the params?

File: framework/src/modules/token/cc_commands/cc_transfer.ts (hunk 4 above)
Comment (shuse2) [reply framework/src/modules/token/cc_commands/cc_transfer.ts:76]: i think you need to rebase to the feature branch

Comment (mosmartin) [reply framework/src/modules/token/cc_commands/cc_transfer.ts:76]: Cool! 😎

**Comment by mosmartin** [comment]
Moved this draft pr to https://github.com/LiskHQ/lisk-sdk/pull/7641

**Pull request closed without merging**